diff --git a/service/crypto.py b/service/crypto.py
index 35c67e5..2799c23 100644
--- a/service/crypto.py
+++ b/service/crypto.py
@@ -56,7 +56,7 @@ class DSAKey:
         return DSAPubKey(self.p, self.q, self.g, self.y)
 
     def sign(self, msg):
-        k = H(self.y)
+        k = number.getRandomRange(1, self.q)
         r = pow(self.g, k, self.p) % self.q
         s = inverse(k, self.q) * (H(msg) + r * self.x) % self.q
         return r, s
@@ -82,6 +82,8 @@ class DSAPubKey:
 
     def verify(self, msg, signature):
         r, s = signature
+        if r == 0 and (s == 1 or s % self.q == 0):
+            return False
         w = inverse(s, self.q)
         u1 = H(msg) * w % self.q
         u2 = r * w % self.q
